File: tarefas-compartilhadas/src/main/java/br/edu/tarefascompartilhadas/service/AmizadeService.java
package br.edu.tarefascompartilhadas.service;

import br.edu.tarefascompartilhadas.enums.StatusAmizade;
import br.edu.tarefascompartilhadas.model.Amizade;
import br.edu.tarefascompartilhadas.model.Usuario;
import br.edu.tarefascompartilhadas.repository.AmizadeRepository;
import br.edu.tarefascompartilhadas.repository.UsuarioRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class AmizadeService {

    private final AmizadeRepository amizadeRepository;
    private final UsuarioRepository usuarioRepository;

    public Amizade criar(Amizade amizade) {
        Long solicitanteId = extrairUsuarioId(
                amizade.getUsuarioSolicitante(),
                "Usuario solicitante e obrigatorio."
        );

        Long amigoId = extrairUsuarioId(
                amizade.getUsuarioAmigo(),
                "Usuario amigo e obrigatorio."
        );

        if (solicitanteId.equals(amigoId)) {
            throw new RuntimeException("Nao e permitido criar amizade do usuario com ele mesmo.");
        }

        if (amizadeRepository.existsByUsuarios(solicitanteId, amigoId)) {
            throw new RuntimeException("Ja existe solicitacao ou amizade entre estes usuarios.");
        }

        Usuario solicitante = buscarUsuario(solicitanteId);
        Usuario amigo = buscarUsuario(amigoId);

        amizade.setId(null);
        amizade.setUsuarioSolicitante(solicitante);
        amizade.setUsuarioAmigo(amigo);
        amizade.setStatusAmizade(StatusAmizade.PENDENTE);
        amizade.setDataSolicitacao(LocalDateTime.now());

        return amizadeRepository.save(amizade);
    }

    public List<Amizade> listar() {
        return amizadeRepository.findAll();
    }

    public Amizade buscarPorId(Long id) {
        return amizadeRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Amizade nao encontrada."));
    }

    public List<Amizade> listarAmigos(Long usuarioId) {
        buscarUsuario(usuarioId);
        return amizadeRepository.findByUsuarioIdAndStatus(usuarioId, StatusAmizade.ACEITA);
    }

    public List<Amizade> listarSolicitacoes(Long usuarioId) {
        buscarUsuario(usuarioId);
        return amizadeRepository.findByUsuarioAmigoIdAndStatusAmizade(
                usuarioId,
                StatusAmizade.PENDENTE
        );
    }

    public Amizade aceitar(Long id, Long usuarioLogadoId) {
        Amizade amizade = buscarPorId(id);

        validarUsuarioPodeResponder(amizade, usuarioLogadoId);

        if (amizade.getStatusAmizade() != StatusAmizade.PENDENTE) {
            throw new RuntimeException("Apenas solicitacoes pendentes podem ser aceitas.");
        }

        amizade.setStatusAmizade(StatusAmizade.ACEITA);
        return amizadeRepository.save(amizade);
    }

    public Amizade recusar(Long id, Long usuarioLogadoId) {
        Amizade amizade = buscarPorId(id);

        validarUsuarioPodeResponder(amizade, usuarioLogadoId);

        if (amizade.getStatusAmizade() != StatusAmizade.PENDENTE) {
            throw new RuntimeException("Apenas solicitacoes pendentes podem ser recusadas.");
        }

        amizade.setStatusAmizade(StatusAmizade.RECUSADA);
        return amizadeRepository.save(amizade);
    }

    public void deletar(Long id) {
        Amizade amizade = buscarPorId(id);
        amizadeRepository.delete(amizade);
    }

    private void validarUsuarioPodeResponder(Amizade amizade, Long usuarioLogadoId) {
        if (usuarioLogadoId == null) {
            throw new RuntimeException("Usuario logado e obrigatorio para responder a solicitacao.");
        }

        buscarUsuario(usuarioLogadoId);

        if (amizade.getUsuarioAmigo() == null || amizade.getUsuarioAmigo().getId() == null) {
            throw new RuntimeException("Nao foi possivel identificar o destinatario da solicitacao.");
        }

        if (!amizade.getUsuarioAmigo().getId().equals(usuarioLogadoId)) {
            throw new RuntimeException("Apenas o destinatario da solicitacao pode aceitar ou recusar.");
        }
    }

    private Usuario buscarUsuario(Long id) {
        return usuarioRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Usuario nao encontrado."));
    }

    private Long extrairUsuarioId(Usuario usuario, String mensagemErro) {
        if (usuario == null || usuario.getId() == null) {
            throw new RuntimeException(mensagemErro);
        }

        return usuario.getId();
    }
}

File: app_tarefas_compartilhadas/lib/services/amizade_service.dart
import '../models/amizade.dart';
import 'api_service.dart';

class AmizadeService {
  final ApiService _api = ApiService();

  Future<List<Amizade>> listar() async {
    final lista = await _api.getList('/amizades');
    return lista.map((item) => Amizade.fromJson(item)).toList();
  }

  Future<Amizade> buscarPorId(int id) async {
    final item = await _api.getObject('/amizades/$id');
    return Amizade.fromJson(item);
  }

  Future<List<Amizade>> listarAmigos(int usuarioId) async {
    final lista = await _api.getList('/usuarios/$usuarioId/amigos');
    return lista.map((item) => Amizade.fromJson(item)).toList();
  }

  Future<List<Amizade>> listarSolicitacoes(int usuarioId) async {
    final lista = await _api.getList('/usuarios/$usuarioId/solicitacoes');
    return lista.map((item) => Amizade.fromJson(item)).toList();
  }

  Future<Amizade> criar(Amizade amizade) async {
    final item = await _api.post('/amizades', amizade.toJson());
    return Amizade.fromJson(item);
  }

  Future<Amizade> aceitar({
    required int id,
    required int usuarioLogadoId,
  }) async {
    final item = await _api.put('/amizades/$id/aceitar', {
      'usuarioLogadoId': usuarioLogadoId,
    });

    return Amizade.fromJson(item);
  }

  Future<Amizade> recusar({
    required int id,
    required int usuarioLogadoId,
  }) async {
    final item = await _api.put('/amizades/$id/recusar', {
      'usuarioLogadoId': usuarioLogadoId,
    });

    return Amizade.fromJson(item);
  }

  Future<void> deletar(int id) async {
    await _api.delete('/amizades/$id');
  }
}